### PR TITLE
Fully support the "presentation layer design"

### PR DESCRIPTION
**Pull request by taiki45**
Convert to resource before checking permissions.

We often create a presentation layer to define Garage resource
definition instead of defining resource definition in the AR model
directly. In that case, Garage automatically converts AR models
to resource objects by calling `to_resource` method on them, but Garage
does not convert AR models to resource objects on checking permissions.
By this change, patching Garage's RestfulActions or converting to
resource object by hand become unnecessary.

### CODE REVIEW

File: lib/garage/restful_actions.rb [hunk 1]
```diff
@@ -32,7 +32,22 @@ def resource_class=(klass)
...
+        @resource_class ||= resource_class_name.constantize
+      end
+
+      private
+
+      def resource_class_name
+        class_name = name.sub(/Controller\z/, '').demodulize.singularize
+        class_name_with_suffix = "#{class_name}Resource"
+        case
+        when const_defined?(class_name_with_suffix)
```
Comment (moro) [inline lib/garage/restful_actions.rb]: Does it work under `Rails.env.development?` (or other lazy-loading) environment?

Comment (moro) [inline lib/garage/restful_actions.rb]: (YAGNI?) I may want to customize naming convention not `FooResource`  eg. `FooPresentation`.

I propose to extract this resource class resolver like this:

```ruby
class Garage::ResourceClassResolver
  def initialize(&block)
    @block = block
  end

  def call(controller_class_name)
    class_name = class_name.sub(/Controller\z/, '').demodulize.singularize
    
    if const_defined?(@blocl.call(class_name)
      ...
    elsif const_defined?(class_name)
      ...
    else
      ...
    end
  end
end

# defalut initalization
Garage.configuration.resouce_class_resolver =
  Garage::ResourceClassResolver.new {|class_name| "#{class_name}Resource" }

# in RestfulActions
def resource_class
  Garage.configuration.resouce_class_resolver.call(name)
end
```
Then I can customize lookup pattern by setting `Garage.configuration.resouce_class_resolver =` in my `config/initializers` .

How do you think about this?

**Comment by moro** [comment]
Concept looks so good to me.

My comment about extraction may be YAGNI. Pick it if you like or ignore it.

File: lib/garage/restful_actions.rb (hunk 1 above)
Comment (taiki45) [inline lib/garage/restful_actions.rb]: I forgot about lazy-loading....

Comment (taiki45) [inline lib/garage/restful_actions.rb]: We can add custome name convention after we need it.

Comment (moro) [inline lib/garage/restful_actions.rb]: 🙆

**Comment by taiki45** [comment]
55e37b3 to avoid auto-loading problems.

**Pull request merged**